In [1]:
%load_ext autoreload
%autoreload 2

In [38]:

from XOR import Population, Genome, Connection, Specie
Specie.threshold

3

In [54]:
INPUT = [[0,0], [0,1], [1,0], [1,1]]
OUTPUT = [0, 1, 1, 0]
p = Population(50, INPUT, OUTPUT, (2,1) )
p.run()

Generation 0 Best fitness 0
Specie 0, pop:$51, avg f0.078 has 50 offspring when total av is $0.078
generation 0 had 0 zeroed genomes
Generation 1 Best fitness 4.0
Specie 0, pop:$47, avg f0.085 has 3 offspring when total av is $1.418
Specie 1, pop:$3, avg f1.333 has 47 offspring when total av is $1.418
generation 1 had 0 zeroed genomes
Generation 2 Best fitness 4.0
Specie 0, pop:$5, avg f0.8 has 13 offspring when total av is $2.893
Specie 1, pop:$43, avg f0.093 has 1 offspring when total av is $2.893
Specie 2, pop:$2, avg f2.0 has 34 offspring when total av is $2.893
generation 2 had 0 zeroed genomes
Generation 3 Best fitness 4.0
Specie 0, pop:$15, avg f0.267 has 3 offspring when total av is $4.388
Specie 1, pop:$1, avg f4.0 has 45 offspring when total av is $4.388
Specie 2, pop:$33, avg f0.121 has 1 offspring when total av is $4.388
generation 3 had 0 zeroed genomes
Generation 4 Best fitness 4.0
Specie 0, pop:$5, avg f0.8 has 13 offspring when total av is $2.893
Specie 1, pop:$43, avg 

In [40]:
top5 = sorted(p.organisms, key=lambda x:x.fitness, reverse=True)[:5]
for o in top5:
    o.show()
    print(o)
    !open example.html

total_conns = 0
total_nodes = 0
for x in p.organisms:
    total_conns += len(x.connections)
    total_nodes += len(x.nodes)

average_nodes = total_nodes/len(p.organisms)
average_conns = total_conns/len(p.organisms)
print("average nodes: ", average_nodes )
print("average connections: ",total_conns/len(p.organisms))

example.html
Genome f:$6.25, nodes:3, conns:2
example.html
Genome f:$6.25, nodes:3, conns:2
example.html
Genome f:$6.25, nodes:3, conns:2
example.html
Genome f:$6.25, nodes:3, conns:2
example.html
Genome f:$6.25, nodes:3, conns:2
average nodes:  5.04
average connections:  6.38


In [22]:
bigger = [x for x in p.organisms if len(x.connections) > average_conns or len(x.nodes) > average_nodes]
for b in bigger:
    print(b)
    b.show()
    !open example.html





Genome f:$6.24, nodes:6, conns:9
example.html
Genome f:$6.24, nodes:5, conns:7
example.html
Genome f:$6.24, nodes:5, conns:7
example.html
Genome f:$6.24, nodes:5, conns:7
example.html
Genome f:$6.24, nodes:5, conns:7
example.html
Genome f:$4.29, nodes:5, conns:7
example.html
Genome f:$4.29, nodes:5, conns:7
example.html
Genome f:$4.29, nodes:5, conns:7
example.html
Genome f:$4.29, nodes:6, conns:9
example.html
Genome f:$4.29, nodes:5, conns:7
example.html
Genome f:$4.29, nodes:5, conns:7
example.html
Genome f:$4.29, nodes:5, conns:7
example.html
Genome f:$4.29, nodes:5, conns:7
example.html
Genome f:$4.29, nodes:5, conns:8
example.html
Genome f:$4.29, nodes:5, conns:7
example.html
Genome f:$4.29, nodes:5, conns:7
example.html
Genome f:$4.29, nodes:5, conns:7
example.html
Genome f:$4.29, nodes:5, conns:7
example.html
Genome f:$4.29, nodes:5, conns:7
example.html
Genome f:$4.29, nodes:5, conns:7
example.html
Genome f:$4.29, nodes:5, conns:7
example.html
Genome f:$4.29, nodes:5, conns:7
e

In [51]:


for o in p.species[-1].organisms: print(o.fitness, o.adjusted_fitness)




4.0 4.0
4.0 4.0
4.0 4.0
4.0 4.0
4.0 4.0
4.0 4.0
4.0 4.0
4.0 4.0
4.0 4.0
4.0 4.0
4.0 4.0
4.0 4.0
4.0 4.0
4.0 4.0
4.0 4.0
4.0 4.0
4.0 4.0
4.0 4.0
4.0 4.0
4.0 4.0
4.0 4.0
4.0 4.0
4.0 4.0
4.0 4.0
4.0 4.0
4.0 4.0
4.0 4.0
4.0 4.0
4.0 4.0


In [41]:

p.speciate()

In [10]:
for o in p.species[1].organisms:
    o.show()
    !open example.html

example.html
example.html
example.html
example.html
example.html
example.html
example.html
example.html
example.html
example.html
example.html
example.html
example.html
example.html
example.html
example.html
example.html
example.html
example.html


In [92]:


[s for s in p.species if len([b for b in bigger if b in s.organisms]) > 0]



[]

In [7]:
p.speciate()

In [9]:
Connection.innov_table

{(0, 2): 0,
 (1, 2): 1,
 (1, 3): 2,
 (3, 2): 3,
 (0, 3): 4,
 (0, 4): 5,
 (4, 2): 6,
 (1, 4): 7,
 (4, 3): 8,
 (3, 4): 9,
 (3, 5): 10,
 (5, 2): 11,
 (4, 5): 12,
 (5, 3): 13,
 (0, 6): 14,
 (6, 4): 15,
 (1, 5): 16,
 (3, 6): 17,
 (6, 5): 18,
 (5, 6): 19,
 (6, 2): 20,
 (0, 5): 21,
 (4, 6): 22,
 (6, 3): 23,
 (1, 6): 24,
 (5, 4): 25,
 (1, 7): 26,
 (7, 2): 27,
 (7, 3): 28,
 (3, 7): 29,
 (7, 5): 30,
 (0, 7): 31,
 (7, 4): 32,
 (0, 8): 33,
 (8, 7): 34,
 (7, 6): 35,
 (5, 8): 36,
 (8, 2): 37,
 (7, 8): 38,
 (4, 7): 39,
 (4, 8): 40,
 (1, 9): 41,
 (9, 5): 42,
 (6, 9): 43,
 (4, 10): 44,
 (10, 3): 45,
 (6, 10): 46,
 (10, 9): 47,
 (0, 9): 48,
 (5, 7): 49,
 (6, 7): 50,
 (8, 3): 51,
 (1, 8): 52,
 (8, 4): 53,
 (8, 6): 54,
 (6, 8): 55,
 (8, 5): 56,
 (3, 8): 57,
 (3, 9): 58,
 (9, 2): 59,
 (9, 7): 60,
 (4, 9): 61,
 (9, 3): 62,
 (9, 4): 63,
 (7, 9): 64}

In [8]:
Specie.threshold

3

In [5]:
p.organisms[0].show()
!open example.html

example.html


In [24]:

[o.connections for o in p.organisms]

[[Conn $0, [0 -> 2] enabled, Conn $1, [1 -> 2] enabled],
 [Conn $0, [0 -> 2] enabled, Conn $1, [1 -> 2] enabled],
 [Conn $0, [0 -> 2] enabled, Conn $1, [1 -> 2] enabled],
 [Conn $0, [0 -> 2] enabled, Conn $1, [1 -> 2] enabled],
 [Conn $0, [0 -> 2] enabled, Conn $1, [1 -> 2] enabled],
 [Conn $0, [0 -> 2] enabled, Conn $1, [1 -> 2] enabled],
 [Conn $0, [0 -> 2] enabled,
  Conn $1, [1 -> 2] disabled,
  Conn $2, [1 -> 3] enabled,
  Conn $3, [3 -> 2] enabled],
 [Conn $0, [0 -> 2] enabled, Conn $1, [1 -> 2] enabled],
 [Conn $0, [0 -> 2] enabled, Conn $1, [1 -> 2] enabled],
 [Conn $0, [0 -> 2] enabled, Conn $1, [1 -> 2] enabled],
 [Conn $0, [0 -> 2] enabled, Conn $1, [1 -> 2] enabled],
 [Conn $0, [0 -> 2] enabled, Conn $1, [1 -> 2] enabled],
 [Conn $0, [0 -> 2] enabled, Conn $1, [1 -> 2] enabled],
 [Conn $0, [0 -> 2] enabled, Conn $1, [1 -> 2] enabled],
 [Conn $0, [0 -> 2] enabled, Conn $1, [1 -> 2] enabled],
 [Conn $0, [0 -> 2] enabled,
  Conn $1, [1 -> 2] enabled,
  Conn $4, [0 -> 3] enable

In [2]:
g = Genome(2,1)
g.show()
!open example.html

example.html


In [5]:
g.mutate()

NameError: name 'g' is not defined

In [6]:
g.show()
!open example.html

example.html


In [5]:
g.mutate_add_node()

# Prev idea
I'm gonna combine two ideas:
- only top 20% from every species can reproduce
- crossover can happen between any two genomes, although weights inside species are boosted

First idea ok, second is dogshit

# Current state of affairs:
i can generate a bunch of networks that look like they could work, but for some reason the basic network always gets more fitness than them :// idk why. Also they can't evolve past 6.25?


[Conn $0, [0 -> 2] enabled,
 Conn $1, [1 -> 2] disabled,
 Conn $2, [1 -> 3] enabled,
 Conn $3, [1 -> 3] enabled]